File: icici/src/lib/indicators.ts
/**
 * Shared indicator engine — identical algorithms to Python analysis/indicators.py.
 *
 * All functions use seeded EMA (result[0] = prices[0]) instead of the
 * technicalindicators library, to match Python _ema() exactly.
 *
 * Population std is used for Bollinger (divide by N, matches Python np.std()).
 */

// ── Result types ──────────────────────────────────────────────────────────────

export interface RSIResult {
    header: string;
    period: number;
    overbought: number;
    oversold: number;
    value: number;
    trend: string;
}

export interface EMAResult {
    header: string;
    shortPeriod: number;
    longPeriod: number;
    trend: string;
}

export interface MACDResult {
    header: string;
    shortPeriod: number;
    longPeriod: number;
    signalPeriod: number;
    macd: number;
    signal: number;
    trend: string;
}

export interface BollingerResult {
    header: string;
    period: number;
    numDeviations: number;
    upper: number;
    middle: number;
    lower: number;
    stdDev: number;
    trend: string;
}

export interface ADXResult {
    header: string;
    period: number;
    trend: string;
}

export interface StochasticResult {
    header: string;
    kPeriod: number;
    dPeriod: number;
    trend: string;
}

// ── Name helpers (match Python column names) ──────────────────────────────────

export function rsiName(period: number, ob: number, os: number): string {
    return `RSI_${period}_${ob}_${os}`;
}

export function emaName(short: number, long: number): string {
    return `EMA_${short}_${long}`;
}

export function macdName(s: number, l: number, sig: number): string {
    return `MACD_${s}_${l}_${sig}`;
}

export function bollingerName(period: number, dev: number): string {
    return `Bollinger_${period}_${dev}`;
}

export function adxName(period: number): string {
    return `ADX_${period}`;
}

export function stochasticName(k: number, d: number): string {
    return `Stoch_${k}_${d}`;
}

export function rsiNameReversed(period: number, ob: number, os: number): string {
    return `RSI_${period}_${ob}_${os}_REV`;
}

// ── Core helper ───────────────────────────────────────────────────────────────

/**
 * Seeded EMA — result[0] = prices[0], then exponential smoothing.
 * Matches Python _ema() in indicators.py exactly.
 */
export function seededEMA(prices: number[], period: number): number[] {
    const mult = 2 / (period + 1);
    const result = new Array<number>(prices.length);
    result[0] = prices[0];
    for (let i = 1; i < prices.length; i++) {
        result[i] = (prices[i] - result[i - 1]) * mult + result[i - 1];
    }
    return result;
}

// ── RSI (Wilder smoothing) ────────────────────────────────────────────────────

/**
 * Matches Python rsi_trend() in indicators.py.
 * Manual Wilder smoothing: seed with simple average, then iterate.
 */
export function calcRSI(
    closes: number[],
    period: number,
    overbought: number,
    oversold: number
): RSIResult | null {
    if (closes.length < period + 1) return null;

    const deltas: number[] = [];
    for (let i = 1; i < closes.length; i++) deltas.push(closes[i] - closes[i - 1]);

    const gains  = deltas.map(d => d > 0 ? d : 0);
    const losses = deltas.map(d => d < 0 ? -d : 0);

    let avgGain = gains.slice(0, period).reduce((s, v) => s + v, 0) / period;
    let avgLoss = losses.slice(0, period).reduce((s, v) => s + v, 0) / period;

    for (let i = period; i < gains.length; i++) {
        avgGain = (avgGain * (period - 1) + gains[i]) / period;
        avgLoss = (avgLoss * (period - 1) + losses[i]) / period;
    }

    const value = avgLoss === 0 ? 100 : 100 - (100 / (1 + avgGain / avgLoss));
    const trend = value > overbought ? 'DOWN' : value < oversold ? 'UP' : 'NEUTRAL';

    return { header: rsiName(period, overbought, oversold), period, overbought, oversold, value, trend };
}

// ── RSI Reversed (contrarian) ─────────────────────────────────────────────────

/**
 * Matches Python rsi_trend_reversed() in indicators.py.
 * Uses simple mean RSI (NOT Wilder smoothing) over the last `period` bars.
 * Reversed interpretation: overbought → UP (buy strength), oversold → DOWN.
 */
export function calcRSIReversed(
    closes: number[],
    period: number,
    overbought: number,
    oversold: number
): RSIResult | null {
    if (closes.length < period + 1) return null;

    // Simple mean over last `period` deltas (matches Python rsi_trend_reversed)
    let totalGain = 0, totalLoss = 0;
    for (let i = closes.length - period; i < closes.length; i++) {
        const delta = closes[i] - closes[i - 1];
        if (delta > 0) totalGain += delta;
        else            totalLoss += -delta;
    }

    const avgGain = totalGain / period;
    const avgLoss = totalLoss / period;

    // No losses = strong uptrend → contrarian buy signal
    if (avgLoss === 0) {
        return { header: rsiNameReversed(period, overbought, oversold), period, overbought, oversold, value: 100, trend: 'UP' };
    }

    const value = 100 - (100 / (1 + avgGain / avgLoss));
    // REVERSED: overbought → UP, oversold → DOWN (contrarian)
    const trend = value > overbought ? 'UP' : value < oversold ? 'DOWN' : 'NEUTRAL';

    return { header: rsiNameReversed(period, overbought, oversold), period, overbought, oversold, value, trend };
}

// ── EMA Crossover ─────────────────────────────────────────────────────────────

/**
 * Matches Python ema_crossover_trend() in indicators.py.
 * Slices to longPeriod, scans for first crossover event.
 */
export function calcEMACrossover(
    closes: number[],
    shortPeriod: number,
    longPeriod: number
): EMAResult | null {
    if (closes.length < longPeriod) return null;

    const prices = closes.slice(-longPeriod);
    const shortEMA = seededEMA(prices, shortPeriod);
    const longEMA  = seededEMA(prices, longPeriod);

    let trend = 'NEUTRAL';
    for (let i = 1; i < prices.length; i++) {
        if (shortEMA[i - 1] <= longEMA[i - 1] && shortEMA[i] > longEMA[i]) {
            trend = 'UP';
            break;
        }
        if (shortEMA[i - 1] >= longEMA[i - 1] && shortEMA[i] < longEMA[i]) {
            trend = 'DOWN';
            break;
        }
    }

    return { header: emaName(shortPeriod, longPeriod), shortPeriod, longPeriod, trend };
}

// ── MACD ──────────────────────────────────────────────────────────────────────

/**
 * Matches Python macd_trend() in indicators.py.
 * Uses seeded EMA on the SAME array (totalNeeded prices), element-wise subtraction.
 */
export function calcMACD(
    closes: number[],
    shortPeriod: number,
    longPeriod: number,
    signalPeriod: number
): MACDResult | null {
    const totalNeeded = signalPeriod + longPeriod;
    if (closes.length < totalNeeded) return null;

    const prices    = closes.slice(-totalNeeded);
    const shortEMA  = seededEMA(prices, shortPeriod);
    const longEMA   = seededEMA(prices, longPeriod);
    const macdLine  = shortEMA.map((v, i) => v - longEMA[i]);
    const signalLine = seededEMA(macdLine, signalPeriod);

    const macd   = macdLine[macdLine.length - 1];
    const signal = signalLine[signalLine.length - 1];
    const trend  = macd > signal ? 'UP' : 'DOWN';

    return {
        header: macdName(shortPeriod, longPeriod, signalPeriod),
        shortPeriod, longPeriod, signalPeriod,
        macd: Math.round(macd * 100) / 100,
        signal: Math.round(signal * 100) / 100,
        trend,
    };
}

// ── Bollinger Bands ───────────────────────────────────────────────────────────

/**
 * Matches Python bollinger_trend() in indicators.py.
 * Population std (divide by N, not N-1) to match Python np.std().
 */
export function calcBollinger(
    closes: number[],
    period: number,
    numDeviations: number
): BollingerResult | null {
    if (closes.length < period) return null;

    const prices  = closes.slice(-period);
    const sma     = prices.reduce((s, p) => s + p, 0) / period;
    const variance = prices.reduce((s, p) => s + Math.pow(p - sma, 2), 0) / period;
    const stdDev  = Math.round(Math.sqrt(variance) * 100) / 100;
    const upper   = Math.round((sma + stdDev * numDeviations) * 100) / 100;
    const lower   = Math.round((sma - stdDev * numDeviations) * 100) / 100;
    const middle  = Math.round(sma * 100) / 100;
    const latest  = prices[prices.length - 1];
    const trend   = latest > upper ? 'UP' : latest < lower ? 'DOWN' : 'NEUTRAL';

    return {
        header: bollingerName(period, numDeviations),
        period, numDeviations,
        upper, middle, lower, stdDev,
        trend,
    };
}

// ── ADX ───────────────────────────────────────────────────────────────────────

/**
 * Matches Python adx_trend() in indicators.py.
 * Wilder-smoothed ATR, +DM, -DM; ADX = smoothed DX.
 */
export function calcADX(
    highs: number[],
    lows: number[],
    closes: number[],
    period: number
): ADXResult | null {
    const n = closes.length;
    if (n < period * 2) return null;

    const tr: number[]      = [];
    const plusDM: number[]  = [];
    const minusDM: number[] = [];

    for (let i = 1; i < n; i++) {
        tr.push(Math.max(
            highs[i] - lows[i],
            Math.abs(highs[i] - closes[i - 1]),
            Math.abs(lows[i] - closes[i - 1])
        ));
        const upMove   = highs[i] - highs[i - 1];
        const downMove = lows[i - 1] - lows[i];
        plusDM.push(upMove > downMove && upMove > 0 ? upMove : 0);
        minusDM.push(downMove > upMove && downMove > 0 ? downMove : 0);
    }

    const len = tr.length;
    const atr    = new Array<number>(len).fill(0);
    const sPlus  = new Array<number>(len).fill(0);
    const sMinus = new Array<number>(len).fill(0);

    atr[period - 1]    = tr.slice(0, period).reduce((s, v) => s + v, 0) / period;
    sPlus[period - 1]  = plusDM.slice(0, period).reduce((s, v) => s + v, 0) / period;
    sMinus[period - 1] = minusDM.slice(0, period).reduce((s, v) => s + v, 0) / period;

    for (let i = period; i < len; i++) {
        atr[i]    = (atr[i - 1] * (period - 1) + tr[i]) / period;
        sPlus[i]  = (sPlus[i - 1] * (period - 1) + plusDM[i]) / period;
        sMinus[i] = (sMinus[i - 1] * (period - 1) + minusDM[i]) / period;
    }

    const plusDI  = atr.map((a, i) => a !== 0 ? 100 * sPlus[i] / a : 0);
    const minusDI = atr.map((a, i) => a !== 0 ? 100 * sMinus[i] / a : 0);
    const dx      = plusDI.map((p, i) => {
        const sum = p + minusDI[i];
        return sum !== 0 ? 100 * Math.abs(p - minusDI[i]) / sum : 0;
    });

    const adxVals = new Array<number>(dx.length).fill(0);
    const start   = 2 * period - 1;
    if (start >= dx.length) return null;

    adxVals[start] = dx.slice(period, start + 1).reduce((s, v) => s + v, 0) / (start - period + 1);
    for (let i = start + 1; i < dx.length; i++) {
        adxVals[i] = (adxVals[i - 1] * (period - 1) + dx[i]) / period;
    }

    const latestADX   = adxVals[adxVals.length - 1];
    const latestPlus  = plusDI[plusDI.length - 1];
    const latestMinus = minusDI[minusDI.length - 1];

    let trend: string;
    if (latestADX < 20)              trend = 'NEUTRAL';
    else if (latestPlus > latestMinus) trend = 'UP';
    else                               trend = 'DOWN';

    return { header: adxName(period), period, trend };
}

// ── Stochastic ────────────────────────────────────────────────────────────────

/**
 * Matches Python stochastic_trend() in indicators.py.
 * %K = (close - lowest) / (highest - lowest) * 100; %D = SMA of %K.
 */
export function calcStochastic(
    highs: number[],
    lows: number[],
    closes: number[],
    kPeriod: number,
    dPeriod: number
): StochasticResult | null {
    const n = closes.length;
    if (n < kPeriod + dPeriod) return null;

    const kValues: number[] = [];
    for (let i = kPeriod - 1; i < n; i++) {
        const highest = Math.max(...highs.slice(i - kPeriod + 1, i + 1));
        const lowest  = Math.min(...lows.slice(i - kPeriod + 1, i + 1));
        kValues.push(highest === lowest ? 50 : 100 * (closes[i] - lowest) / (highest - lowest));
    }

    const dValues: number[] = [];
    for (let i = dPeriod - 1; i < kValues.length; i++) {
        dValues.push(kValues.slice(i - dPeriod + 1, i + 1).reduce((s, v) => s + v, 0) / dPeriod);
    }

    const latestK = kValues[kValues.length - 1];
    const latestD = dValues.length > 0 ? dValues[dValues.length - 1] : latestK;

    let trend: string;
    if (latestK < 20 && latestK > latestD)      trend = 'UP';
    else if (latestK > 80 && latestK < latestD) trend = 'DOWN';
    else                                          trend = 'NEUTRAL';

    return { header: stochasticName(kPeriod, dPeriod), kPeriod, dPeriod, trend };
}

File: icici/analysis/indicators.py
import warnings
import numpy as np

warnings.filterwarnings("ignore", category=RuntimeWarning)


def _round2(n: float) -> float:
    return round(n, 2)


# ---------------------------------------------------------------------------
# RSI  (mirrors decision.ts _calculateRSITrend lines 927-954)
# ---------------------------------------------------------------------------

def rsi_trend(closing_prices: list[float], period: int, overbought: int, oversold: int) -> str | None:
    if len(closing_prices) < period + 1:
        return None

    prices = np.array(closing_prices, dtype=float)
    deltas = np.diff(prices)

    gains = np.where(deltas > 0, deltas, 0.0)
    losses = np.where(deltas < 0, -deltas, 0.0)

    avg_gain = np.mean(gains[:period])
    avg_loss = np.mean(losses[:period])

    for i in range(period, len(gains)):
        avg_gain = (avg_gain * (period - 1) + gains[i]) / period
        avg_loss = (avg_loss * (period - 1) + losses[i]) / period

    if avg_loss == 0:
        latest_rsi = 100.0
    else:
        rs = avg_gain / avg_loss
        latest_rsi = 100 - (100 / (1 + rs))

    if latest_rsi > overbought:
        return "DOWN"
    elif latest_rsi < oversold:
        return "UP"
    return "NEUTRAL"


def rsi_column_name(period, overbought, oversold) -> str:
    return f"RSI_{period}_{overbought}_{oversold}"


# ---------------------------------------------------------------------------
# MACD  (mirrors decision.ts _calculateMACDTrend lines 887-925)
# ---------------------------------------------------------------------------

def _ema(values: np.ndarray, period: int) -> np.ndarray:
    result = np.empty(len(values))
    multiplier = 2 / (period + 1)
    result[0] = values[0]
    for i in range(1, len(values)):
        result[i] = (values[i] - result[i - 1]) * multiplier + result[i - 1]
    return result


def macd_trend(closing_prices: list[float], short_period: int, long_period: int, signal_period: int) -> str | None:
    total_needed = signal_period + long_period
    if len(closing_prices) < total_needed:
        return None

    prices = np.array(closing_prices[-total_needed:], dtype=float)
    short_ema = _ema(prices, short_period)
    long_ema = _ema(prices, long_period)

    # MACD line: short EMA - long EMA (both full length)
    macd_line = short_ema - long_ema
    if len(macd_line) < signal_period:
        return None

    signal_line = _ema(macd_line, signal_period)

    latest_macd = macd_line[-1]
    latest_signal = signal_line[-1]

    return "UP" if latest_macd > latest_signal else "DOWN"


def macd_column_name(short, long, signal) -> str:
    return f"MACD_{short}_{long}_{signal}"


# ---------------------------------------------------------------------------
# Bollinger Bands  (mirrors decision.ts _calculateBollingerBandsTrend lines 956-990)
# ---------------------------------------------------------------------------

def bollinger_trend(closing_prices: list[float], period: int, num_deviations: float) -> str | None:
    if len(closing_prices) < period:
        return None

    prices = np.array(closing_prices[-period:], dtype=float)
    sma = np.mean(prices)
    # population std (matches TypeScript: divides by N, not N-1)
    std_dev = np.std(prices)

    upper_band = sma + (std_dev * num_deviations)
    lower_band = sma - (std_dev * num_deviations)
    latest = prices[-1]

    if latest > upper_band:
        return "UP"
    elif latest < lower_band:
        return "DOWN"
    return "NEUTRAL"


def bollinger_column_name(period, num_deviations) -> str:
    return f"Bollinger_{period}_{num_deviations}"


# ---------------------------------------------------------------------------
# EMA Crossover  (mirrors decision.ts _detectEMATrend lines 857-885)
# ---------------------------------------------------------------------------

def ema_crossover_trend(closing_prices: list[float], short_period: int, long_period: int) -> str | None:
    if len(closing_prices) < long_period:
        return None

    prices = np.array(closing_prices[-long_period:], dtype=float)
    short_ema = _ema(prices, short_period)
    long_ema = _ema(prices, long_period)

    for i in range(1, len(prices)):
        if short_ema[i - 1] <= long_ema[i - 1] and short_ema[i] > long_ema[i]:
            return "UP"
        elif short_ema[i - 1] >= long_ema[i - 1] and short_ema[i] < long_ema[i]:
            return "DOWN"

    return "NEUTRAL"


def ema_column_name(short, long) -> str:
    return f"EMA_{short}_{long}"


# ---------------------------------------------------------------------------
# ADX  (new indicator — Average Directional Index)
# ---------------------------------------------------------------------------

def adx_trend(highs: list[float], lows: list[float], closes: list[float], period: int = 14) -> str | None:
    n = len(closes)
    if n < period * 2:
        return None

    h = np.array(highs, dtype=float)
    lo = np.array(lows, dtype=float)
    c = np.array(closes, dtype=float)

    # True Range
    tr = np.maximum(h[1:] - lo[1:], np.maximum(np.abs(h[1:] - c[:-1]), np.abs(lo[1:] - c[:-1])))
    # +DM / -DM
    plus_dm = np.where((h[1:] - h[:-1]) > (lo[:-1] - lo[1:]), np.maximum(h[1:] - h[:-1], 0), 0)
    minus_dm = np.where((lo[:-1] - lo[1:]) > (h[1:] - h[:-1]), np.maximum(lo[:-1] - lo[1:], 0), 0)

    # Smoothed with Wilder's method
    atr = np.zeros(len(tr))
    s_plus = np.zeros(len(tr))
    s_minus = np.zeros(len(tr))
    atr[period - 1] = np.mean(tr[:period])
    s_plus[period - 1] = np.mean(plus_dm[:period])
    s_minus[period - 1] = np.mean(minus_dm[:period])

    for i in range(period, len(tr)):
        atr[i] = (atr[i - 1] * (period - 1) + tr[i]) / period
        s_plus[i] = (s_plus[i - 1] * (period - 1) + plus_dm[i]) / period
        s_minus[i] = (s_minus[i - 1] * (period - 1) + minus_dm[i]) / period

    plus_di = np.where(atr != 0, 100 * s_plus / atr, 0)
    minus_di = np.where(atr != 0, 100 * s_minus / atr, 0)
    dx = np.where((plus_di + minus_di) != 0, 100 * np.abs(plus_di - minus_di) / (plus_di + minus_di), 0)

    # ADX = smoothed DX
    adx_vals = np.zeros(len(dx))
    start = 2 * period - 1
    if start >= len(dx):
        return None
    adx_vals[start] = np.mean(dx[period:start + 1])
    for i in range(start + 1, len(dx)):
        adx_vals[i] = (adx_vals[i - 1] * (period - 1) + dx[i]) / period

    latest_adx = adx_vals[-1]
    latest_plus = plus_di[-1]
    latest_minus = minus_di[-1]

    if latest_adx < 20:
        return "NEUTRAL"
    elif latest_plus > latest_minus:
        return "UP"
    else:
        return "DOWN"


def adx_column_name(period) -> str:
    return f"ADX_{period}"


# ---------------------------------------------------------------------------
# Stochastic Oscillator  (new indicator)
# ---------------------------------------------------------------------------

def stochastic_trend(highs: list[float], lows: list[float], closes: list[float],
                     k_period: int = 14, d_period: int = 3) -> str | None:
    n = len(closes)
    if n < k_period + d_period:
        return None

    h = np.array(highs, dtype=float)
    lo = np.array(lows, dtype=float)
    c = np.array(closes, dtype=float)

    # %K
    k_values = []
    for i in range(k_period - 1, n):
        highest = np.max(h[i - k_period + 1:i + 1])
        lowest = np.min(lo[i - k_period + 1:i + 1])
        if highest == lowest:
            k_values.append(50.0)
        else:
            k_values.append(100 * (c[i] - lowest) / (highest - lowest))

    k_arr = np.array(k_values)
    # %D = SMA of %K
    d_values = []
    for i in range(d_period - 1, len(k_arr)):
        d_values.append(np.mean(k_arr[i - d_period + 1:i + 1]))

    latest_k = k_arr[-1]
    latest_d = d_values[-1] if d_values else latest_k

    # Overbought > 80, oversold < 20
    if latest_k < 20 and latest_k > latest_d:
        return "UP"
    elif latest_k > 80 and latest_k < latest_d:
        return "DOWN"
    return "NEUTRAL"


def stochastic_column_name(k_period, d_period) -> str:
    return f"Stoch_{k_period}_{d_period}"


# ---------------------------------------------------------------------------
# Reversed RSI (Contrarian Strategy)
# ---------------------------------------------------------------------------

def rsi_trend_reversed(closing_prices: list, period: int, overbought: int, oversold: int) -> str:
    """
    Reversed RSI logic for contrarian strategies.
    Overbought (>70) → UP signal (buy the strength)
    Oversold (<30) → DOWN signal (sell the weakness)
    """
    if len(closing_prices) < period + 1:
        return "NEUTRAL"

    # Calculate RSI (same as normal)
    gains = []
    losses = []
    for i in range(len(closing_prices) - period, len(closing_prices)):
        delta = closing_prices[i] - closing_prices[i - 1]
        if delta > 0:
            gains.append(delta)
            losses.append(0)
        else:
            gains.append(0)
            losses.append(abs(delta))

    avg_gain = np.mean(gains)
    avg_loss = np.mean(losses)

    if avg_loss == 0:
        return "UP"  # No losses = strong uptrend

    rs = avg_gain / avg_loss
    rsi = 100 - (100 / (1 + rs))

    # REVERSED interpretation
    if rsi > overbought:
        return "UP"  # Contrarian: overbought = buy signal
    elif rsi < oversold:
        return "DOWN"  # Contrarian: oversold = sell signal
    return "NEUTRAL"


def rsi_column_name_reversed(period: int, overbought: int, oversold: int) -> str:
    return f"RSI_{period}_{overbought}_{oversold}_REV"
